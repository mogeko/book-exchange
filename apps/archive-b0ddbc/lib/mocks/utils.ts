import { faker } from "@faker-js/faker";

export function oneOf<T = any>(array: T[]): T {
  return array[randomNum(array.length - 1)];
}

export function someOf<T = any>(array: T[]): T[] {
  const result = array.filter(randomBool);

  return result.length === 0 ? [oneOf(array)] : result;
}

export function languages<T = any>(fn: (lang: string[]) => T): T {
  const langs = [
    "Afrikaans",
    "Arabic",
    "Azerbaijani",
    "Czech",
    "German",
    "Greek",
    "English",
    "Spanish",
    "Farsi",
    "Finnish",
    "French",
    "Georgian",
    "Hebrew",
    "Hrvatski",
    "Hungarian",
    "Armenian",
    "Indonesia",
    "Italian",
    "Japanese",
    "Korean",
    "Latvian",
    "Macedonian",
    "Norwegian",
    "Nepalese",
    "Dutch",
    "Polish",
    "Romanian",
    "Russian",
    "Slovakian",
    "Swedish",
    "Turkish",
    "Ukrainian",
    "Urdu",
    "Vietnamese",
    "Chinese",
  ];

  return fn(langs);
}

export const randomNum = faker.datatype.number;

export const randomBool = faker.datatype.boolean;

export const randomDateRecent = () => {
  return faker.date.recent(10, "2022-01-01T00:00:00.000Z").toISOString();
};

export const randomDatePost = () => {
  return faker.date.past(1000, "2000-01-01T00:00:00.000Z").toISOString();
};
