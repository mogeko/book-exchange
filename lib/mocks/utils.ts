import { faker } from "@faker-js/faker";

export function oneOf<T = any>(array: T[]): T {
  return array[randomNum(array.length - 1)];
}

export function someOf<T = any>(array: T[]): T[] {
  const result = array.filter(randomBool);

  return result.length === 0 ? [oneOf(array)] : result;
}

export function languages<T = any>(fn: (lang: string[]) => T): T {
  const langs = [
    "Afrikaans",
    "Arabic",
    "Azerbaijani",
    "Czech",
    "German",
    "Greek",
    "English",
    "Spanish",
    "Farsi",
    "Finnish",
    "French",
    "Georgian",
    "Hebrew",
    "Hrvatski",
    "Hungarian",
    "Armenian",
    "Indonesia",
    "Italian",
    "Japanese",
    "Korean",
    "Latvian",
    "Macedonian",
    "Norwegian",
    "Nepalese",
    "Dutch",
    "Polish",
    "Romanian",
    "Russian",
    "Slovakian",
    "Swedish",
    "Turkish",
    "Ukrainian",
    "Urdu",
    "Vietnamese",
    "Chinese",
  ];

  return fn(langs);
}

export const randomNum = faker.datatype.number;

export const randomBool = faker.datatype.boolean;
