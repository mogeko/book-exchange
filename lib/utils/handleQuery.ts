function handleQuery<T extends ParamProps>(url: URL, params: T) {
  if (Object.keys(params).length === 0) return url;
  const queryParams = Object.entries(params)
    .filter(([_, value]) => value)
    .map(([key, value]) => {
      if (Array.isArray(value)) {
        return value
          .filter(Boolean)
          .map((v) => `${key}=${v}`)
          .join("&");
      }

      return `${key}=${value}`;
    })
    .join("&");

  return [url, queryParams].join("?");
}

interface ParamProps {
  [key: string]: any;
}

export type URL = `/api/${string}`;
export type Query = [URL | undefined, ParamProps];

export default handleQuery;
