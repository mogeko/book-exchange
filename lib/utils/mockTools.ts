import { faker } from "@faker-js/faker";

export function oneOf(array: any[]) {
  return array[randomNum(array.length - 1)];
}

export function randomLanguage() {
  return oneOf([
    "Afrikaans",
    "Arabic",
    "Azerbaijani",
    "Czech",
    "German",
    "Greek",
    "English",
    "Spanish",
    "Farsi",
    "Finnish",
    "French",
    "Georgian",
    "Hebrew",
    "Hrvatski",
    "Hungarian",
    "Armenian",
    "Indonesia",
    "Italian",
    "Japanese",
    "Korean",
    "Latvian",
    "Macedonian",
    "Norwegian",
    "Nepalese",
    "Dutch",
    "Polish",
    "Romanian",
    "Russian",
    "Slovakian",
    "Swedish",
    "Turkish",
    "Ukrainian",
    "Urdu",
    "Vietnamese",
    "Chinese",
  ]);
}

export function arrayBy(length: number, value: any) {
  return Array.from({ length: length }, () => value);
}

export const listBy = arrayBy;

export const randomNum = faker.datatype.number;
