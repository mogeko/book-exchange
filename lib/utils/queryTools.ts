function handleQuery<T extends {}>(url: `/${string}`, params: T) {
  if (Object.keys(params).length === 0) return url;
  const queryParams = Object.entries(params)
    .filter(([_, value]) => value)
    .map(([key, value]) => `${key}=${value}`)
    .join("&");
  return [url, queryParams].join("?");
}

export default handleQuery;
