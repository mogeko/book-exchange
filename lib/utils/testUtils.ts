import { render, RenderOptions } from "@testing-library/react";
import { WrapProvider } from "@/pages/_app";

function renderWithProvider(ui: React.ReactElement, options?: RenderOptions) {
  return render(ui, { wrapper: WrapProvider, ...options });
}

export * from "@testing-library/react";
export { renderWithProvider as render };
