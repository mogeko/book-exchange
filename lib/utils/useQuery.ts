import handleQuery from "@/lib/utils/queryTools";
import useSWR, { type SWRResponse, type SWRConfiguration } from "swr";
import useSWRInfinite, {
  type SWRInfiniteConfiguration,
  type SWRInfiniteResponse,
} from "swr/infinite";

function useQuery<T, U = {}>(
  url: `/${string}`,
  param: U,
  opts: SWRConfiguration<T> = {}
) {
  const query = handleQuery(url, param);
  const res: SWRResponse<T> = useSWR(query, opts);
  const { data, error, ...otherRes } = res;

  return {
    data,
    isLoading: !error && !data,
    isError: error,
    ...otherRes,
  };
}

export function useQueryInfinite<T>(
  getKey: (index: number, previous: T | null) => string | null,
  opts: SWRInfiniteConfiguration<T> = {}
) {
  const res: SWRInfiniteResponse<T> = useSWRInfinite(getKey, opts);
  const { data, error, ...otherRes } = res;

  return {
    data: data,
    isLoading: !error && !data,
    isError: error,
    ...otherRes,
  };
}

export default useQuery;
