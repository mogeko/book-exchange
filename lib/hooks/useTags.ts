import useQuery from "@/lib/hooks/useQuery";
import { type SWRConfiguration } from "swr";

function useTags(opts?: SWRConfiguration<TagsType>) {
  return useQuery<TagsType>("/tags", {}, opts);
}

export interface TagsType {
  [tagsGroup: string]: string[];
}

export default useTags;
