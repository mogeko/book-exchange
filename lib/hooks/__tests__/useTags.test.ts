import "@testing-library/jest-dom";
import { renderHook, waitFor } from "@/lib/utils/testTools";
import useTags from "@/lib/hooks/useTags";

describe("useTags", () => {
  it("should be defined", () => {
    expect(useTags).toBeDefined();
  });

  it("snapshot tags", async () => {
    const { result } = renderHook(() => useTags());

    await waitFor(() => expect(result.current.isLoading).toBe(false));

    expect(result.current).toMatchSnapshot();
  });
});
