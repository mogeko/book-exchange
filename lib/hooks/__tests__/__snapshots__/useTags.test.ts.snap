// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTags snapshot tags 1`] = `
Object {
  "data": Object {
    "Bedfordshire": Array [
      "radical",
      "parse",
      "withdrawal",
      "digital",
      "orchestrate",
      "Table",
      "Account",
      "North",
      "deposit",
    ],
    "Fresh": Array [
      "indigo",
      "orchestrate",
      "portals",
      "Organic",
      "Outdoors",
      "lavender",
      "Card",
      "Turnpike",
      "Oregon",
    ],
    "Kids": Array [
      "utilisation",
      "Oriental",
      "innovative",
      "Pants",
      "Assurance",
    ],
    "Savings": Array [
      "synthesize",
      "Car",
      "Granite",
      "Beauty",
      "Borders",
      "Elegant",
      "Communications",
    ],
    "invoice": Array [
      "Loaf",
      "quantify",
      "Account",
      "Auto",
      "defect",
      "Account",
      "calculating",
      "engineer",
      "Oklahoma",
    ],
  },
  "error": undefined,
  "isLoading": false,
  "isValidating": false,
  "mutate": [Function],
}
`;
